### PR TITLE
v3.0.6 3/3/2016

### LINKED ISSUE #448: OfflineEditAdvanced - after multiple offline restarts, UID begins at -1 again for Adds
Hi Andy, I noticed that after an offline restart, and recreating features from ```offlineFeatureManager.getFeatureLayerJSONDataStore(success,dataStore){...}```, the dataStore correctly populates the OIDs of any features created in the first offline-loaded session with their respective temporary negative numbers.

But on subsequent offline restarts, when creating another new feature, the features created in the new session will start over at -1, which then performs an Update on the previously assigned OID-1, instead of an Add of a new feature with OID -2.

Perhaps I'm just missing something -- Is there a way to point the offline Feature Layer to the existing OIDs of recreated features, so that the OfflineEditAdvanced correctly picks up where the previous session left off?

Thanks,
-rob.

### PR DESCRIPTION
**Pull request by robkspeer**
## Version 3.0.6 - March 3, 2016

No breaking changes

**Bug Fixes**

* Closes #448 - OfflineEditAdvanced - after multiple offline restarts,
UID begins at -1 again for Adds.

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -117,7 +117,7 @@ module.exports = function(grunt) {
                 beautify: {
                     semicolons: false //Required: prevents dojo parser errors w/ minified files in this project
                 },
-                preserveComments: 'some',
```
Comment (andygup) [inline Gruntfile.js]: Hi Rob, please change this to: `preserveComments: /^!/` and then rerun `grunt build`. That should fix the issue.

Apparently the `'some'` option no longer works in the latest versions uglify. I tested the new pattern `v1.0.0` and it worked like a champed.

File: lib/edit/editsStore.js [hunk 2]
```diff
@@ -540,6 +540,58 @@ O.esri.Edit.EditStore = function () {
         }
     };
 
+    /*
+     * Query the database, looking for any existing Add temporary OIDs, and return the nextTempId to be used.
+     * @param feature - extended layer from offline edit advanced
+     * @param callback {int, messageString} or {null, messageString}
+     */
+    this.getNextLowestTempId = function (feature, callback) {
```
Comment (andygup) [inline lib/edit/editsStore.js]: Nicely done!!

File: lib/edit/OfflineEditBasic.js [hunk 3]
```diff
@@ -488,7 +488,19 @@ define([
...
                     // we need to identify ADDs before sending them to the server
                     // we assign temporary ids (using negative numbers to distinguish them from real ids)
-                    layer._nextTempId = -1;
+                    // query the database first to find any existing offline adds, and find the next lowest integer to start with.
+                    this._editStore.getNextLowestTempId(layer, function(value, status){
+                        if(status === "success"){
+                            console.log("_nextTempId:", value);
+                            layer._nextTempId = value;
+                        }
+                        else{
```
Comment (andygup) [inline lib/edit/OfflineEditBasic.js]: When looking at the `else` statement here, I think the only time `status != success` is when there is an error, correct?

If that's true, do we need the `else` statement? 

Also, we could handle errors but that's gonna be messy, messy code that may be overkill.

Thoughts?

**Comment by robkspeer** [comment]
Agreed, the only time the ```else``` would come in is in an error.  And, if the database can't be opened, we're not going to get very far adding anything to it anyway!

I do think a default of ```layer._nextTempId = -1``` needs applied somewhere, just as protection from anything funny happening.

That could happen in the ```else```, or assigning it before calling ```_editStore.getNextLowestTempId()``` as a default.  I just didn't know how the timing of the callback would effect as the code moves along?  I think putting a deferred in there and waiting for it to resolve might be overkill.

Where would you like to see the default of ```-1``` be assigned?  I'll go along with the coding pattern you prefer and rebuild with the proper ```uglify.options.preserveComments: /^!/``` parameter for this PR.

**Comment by andygup** [comment]
>>I do think a default of layer._nextTempId = -1 needs applied somewhere, just as protection from anything funny happening.

I believe that you built that into the database query and it will return -1 by default, right? Reference: https://github.com/Esri/offline-editor-js/pull/450/files#diff-642819f962c2a421135dc243e333877aR574

>>I just didn't know how the timing of the callback would effect as the code moves along?

Should be okay, since it's already wrapped in a deferred. Reference https://github.com/Esri/offline-editor-js/blob/master/lib/edit/OfflineEditBasic.js#L130

**Comment by andygup** [comment]
@robkspeer How's it looking on the pull request?

**Comment by robkspeer** [comment]
Sorry Andy, this got away from me.

For #451, ```test/SpecRunner.offlineTilesAdvanced.TokenBased.html``` shows that the new optional ```dbConfig.offlineIdManager``` parameter can be supplied, or if that line is commented out, the localStorage name will default to "offline_id_manager."

**Comment by andygup** [comment]
Awesome job, thanks for the pull request! 

Just a few steps remaining. I can wrap up pushing the gh-pages update, creating a release and pushing to npm. It's truly no problem on my end, it will only take a 2 - 3 minutes.

**Pull request merged**